text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,103 @@
+
+## Motivation
+
+There are two main motivations for this change.
+
+First, in the status quo, Cedar policy ASTs actually cannot losslessly roundtrip to EST and back.
+This is because the AST representation correctly allows duplicate keys in record literals, but the EST representation drops them (due to an oversight during EST design). In this way, the EST is unfaithful to Cedar semantics, and not all valid Cedar policies are expressible in the EST, which violates the EST's design goal of being lossless modulo whitespace/comments/etc.
+
+More precisely, if a JSON EST contains duplicate record keys in a record literal, our JSON parser will take last-value-wins, and the internal EST representation will contain only the last value.
+This results in the correct behavior on examples like `{ foo: 2, foo: 3 }`, but the incorrect behavior on examples containing errors, like `{ foo: 1 + "hello", foo: 3 }`.
+It also means that roundtripping the AST representation to the EST representation and back necessarily loses information -- specifically, the values associated with all but the last instance of each key.
+
+The second motivation is that the status quo makes things difficult for our Cedar formal specification and accompanying proofs.
+The proofs are required to reason about "well-formed" records as a pre- and post-condition, which could be avoided if records were structurally prohibited from having duplicate keys (e.g., by using a map type instead of an association list).
+The status quo also requires reasoning about the order keys appear in a record literal, meaning that the keys can't be naively sorted without preserving information about their original order.
+If duplicate keys were prohibited, the order of keys wouldn't matter (but see "Unresolved questions" below).
+
+A tertiary motivation for this proposal is that it would bring the Cedar language in line with our desired behavior on duplicate keys in other Cedar API surfaces. See [cedar#159](https://github.com/cedar-policy/cedar/issues/159).
+
+## Detailed design
+
+We'd do something very like [cedar#169](https://github.com/cedar-policy/cedar/pull/169), except that instead of implicitly dropping all but the last value for each key (in `.collect()`), we'd be careful to emit an error when constructing an `Expr` if the record has duplicate keys.
+
+In the case of record literals, the error could be caught in the CST->AST step, resulting in the user seeing a parse error rather than an evaluation error.
+In the case of `context`, the error would need to be caught when constructing the `Context` object.
+In other cases, e.g. if evaluating a Cedar subexpression produces a record with duplicate keys, the user would see an evaluation error. (But see "Unresolved questions" below.)
+
+I don't believe this change would have a significant effect on validation or width/depth subtyping of records, but please correct me if I'm wrong.
+
+## Drawbacks
```

Comment:
I'm supportive of this change as written.  Another mitigating factor is that the current behavior with respect to duplicate keys doesn't seem to be documented.  For example:

* https://docs.cedarpolicy.com/syntax-datatypes.html#record

So, if any policies are relying on this behavior, it's purely by accident.  

Another justification for this change is that records are more closely related to structs than maps in other languages.  With that in mind, allowing duplicate keys during literal construction is counter-intuitive.

Reply:
I couldn't think of any significant effects off the top of my head. That being said, it would make the Dafny type checker (maybe also the evaluator?) a little bit slower since iterating a map in Dafny is not trivial (see: https://github.com/cedar-policy/cedar-spec/blob/main/cedar-dafny/def/util.dfy#L64).

Reply:
@emina, I added text to the RFC re: your comments.

Reply:
Looks good to me!  Approved.